### DISCUSSION TITLE
Trying to implement a computed store feature

### DISCUSSION THREAD

**Discussion by caboodal** [ngrx/platform#4451] [Q&A] [answered]
I've defined the following store feature but when I use this within my store I get compile errors the first one of which is "No overload matches this call." as per usual the error reported is so huge it's very difficult to work out what the root of the problem is. As my store works with the withComputed added as a normal withComputed but then errors with that replaced with the below feature I assume there is something wrong with the code below.

````
import {signalStoreFeature, type, withComputed} from "@ngrx/signals";
import {PagedListResponseState} from "../../../state/paged-list-response-state";
import {computed} from "@angular/core";

export function withResults<T>() {
    return signalStoreFeature(
        { state: type<PagedListResponseState<T>>() },
        withComputed((state) => ({
            results: computed((): Array<T> => {
                return state.response.results();
            })
        }))
    )
}
````

**Comment by caboodal** [comment]
OK it's being caused by the fact that I'm only passing in part of the state type, if I use the complete state type the error goes away i.e. `{ state: type<UserState>() },` but this defeats the object of moving it to a feature so I can share it amongst other list pages. I've not had the same issue with create method store features.

My user state is made up of the following 3 elements, and my withResults feature only needs access to the `PagedListResponseState<T>` portion.

````
export type UserState = {
    filterParams: ListFilterParameters;
} & PagedListQueryState & PagedListResponseState<ListUserResponse>;
````

**Comment by caboodal** [comment]
If I change the feature creation function as below and then call the withResults like this `withResults<ListFundingProviderResponse>()` I get the error but if I call it like this ``withResults()` I get no errors and it works.

````
import {signalStoreFeature, type, withComputed} from "@ngrx/signals";
import {computed} from "@angular/core";
import {PagedListResponseState} from "../../../state/paged-list-response-state";

export function withResults<T>() {
    return signalStoreFeature(
        { state: type<PagedListResponseState<T>>() },
        withComputed(({response}) => ({
            results: computed((): Array<T> => {
                return response.results();
            })
        }))
    )
}
````

**Reply by rainerhahnekamp** [reply]
@caboodal are you running on rc-2?

**Reply by caboodal** [reply]
Hi @rainerhahnekamp 

No I've just checked and I'm using ^18.0.0-rc.1 so I guess I should update.

ncu doesn't show it as an update hence why I've missed it :-/

**Reply by rainerhahnekamp** [reply]
the update isn't shown because it is a release candidate. Let me know if you still have the issue with rc-2 and I'll take a look.

**Reply by caboodal** [reply]
Thank you, I have a number of breaking changes after updating, I'll get those resolved and report back.

**Reply by caboodal** [reply]
OK the compile errors I was getting were related to the two stores I implemented the store feature in. It seems that the store raises errors or doesn't raise errors dependent on the position of the `withResults<ListUserResponse>()` method call. I actually need to move all of these into one store feature but thought I'd mention it.

This doesn't generate errors: -
````
    withApplyFirst<ListUserResponse>(),
    withResults<ListUserResponse>(),
    withApplyFilter(),
    withApplyPageSize(),
    withApplySort(),
````

This does raise errors: -
````
    withResults<ListUserResponse>(),
    withApplyFirst<ListUserResponse>(),
    withApplyFilter(),
    withApplyPageSize(),
    withApplySort(),

````

**Reply by markostanimirovic** [reply]
Hey @caboodal,

Can you please reproduce the issue on StackBlitz or the TypeScript playground?

In the meantime, bump `@ngrx/signals` to the latest version `18.0.0-rc.3` and take a look at this section about known TypeScript issues when creating custom SignalStore features: https://deploy-preview-4436--ngrx-io.netlify.app/guide/signals/signal-store/custom-store-features#known-typescript-issues

**Reply by rainerhahnekamp** [reply]
@caboodal, here you have a StackBlitz which uses already rc.3: https://stackblitz.com/~/github.com/rainerhahnekamp/ngrx-signal-store-starter?file=package.json

**Reply by caboodal** [reply]
I have replicated the issue by mirroring my application as it stands as shown in the Stackblitz below.

If you move the `withResults` method call below the `withApplyFirst` the errors go away.

[Stackblitz](https://stackblitz.com/~/github.com/rainerhahnekamp/ngrx-signal-store-starter?file=package.json)

I have also added the unknown generics to the non-generic features but the problem persists.

**Reply by markostanimirovic** [reply]
@caboodal You sent the link to the starter SignalStore Stackblitz. It's necessary to fork the project on Stackblitz and save the changes.

**Reply by caboodal** [reply]
Sorry, not used Stackblitz before, there's a public git repo here...

[ngrx-store-feature](https://github.com/caboodal/ngrx-store-feature)

**Reply by markostanimirovic** [reply]
Seems like another TS issue to me. 👀 I'm going to investigate this and try to find some workaround.

If you remove explicit generic arguments types will be automatically inferred and it works as expected:

```diff
export const FundingProviderStore = signalStore(
    {providedIn: 'root'},
    withState(initialState),
-   withResults<TestDataItem>(),
-   withApplyFirst<TestDataItem>(),
+   withResults(),
+   withApplyFirst(),
    withApplyFilter(),
    withApplyPageSize(),
    withApplySort(),
    withMethods((state) => ({
        refresh: () => {
            const andy = state.response.pageIndex();
        }
    }))
);
```

By the way, `signalState` shouldn't be used when defining the initial state. Define it as a plain object:

```diff
-const initialState = signalState<TestState>({
+const initialState: TestState = {
  // ...
-});
+};
```

Check the documentation:
- [SignalState guide](https://ngrx.io/guide/signals/signal-state)
- [SignalStore guide](https://ngrx.io/guide/signals/signal-store)

**Reply by caboodal** [reply]
If I can just remove the <TestDataItem> generic type and everything will still work as expected then I guess that's an easier solution and keeps the code cleaner.

Thank you for the heads up on using signal state, I'll review the documentation in the links.

**Reply by caboodal** [reply]
If I remove the generic typing on `withResults` and `withApplyFirst` I get compile errors.

**Reply by markostanimirovic** [reply]
The Angular project you shared is incomplete - it cannot be cloned and run because `package.json`, `angular.json`, etc. are missing.

When I cloned it, I added `package.json`, installed dependencies, and when I removed explicit generic usages error warnings in the IDE disappeared.

**Reply by markostanimirovic** [reply, marked as answer]
Take a look: https://stackblitz.com/edit/github-8kstl6?file=state%2Ftest-store.ts

`angular.json` is missing so the application cannot be started, but when dependencies are installed, error warnings will disappear in `test-store.ts` when explicit generics are removed from `withResults` and `withApplyFirst`.
